[naming-convention] Precedence of property selector has been changed

<!--
If you have a problem with a specific rule, please begin your issue title with [rulename] to make it easier to search for.
I.e. "[no-unused-vars] False positive when fooing the bar"

Please don't ignore this template.

If you ignore it, we're just going to respond asking you to fill it out, which wastes everyone's time.
The more relevant information you can include, the faster we can find the issue and fix it without asking you for more info.
-->

<!--
Are you opening an issue because the rule you're trying to use is not found?
🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨
1) Check the releases log: https://github.com/typescript-eslint/typescript-eslint/releases
    -  If the rule isn't listed there, then chances are it hasn't been released to the main npm tag yet.
2) Try installing the `canary` tag: `npm i @typescript-eslint/eslint-plugin@canary`.
    - The canary tag is built for every commit to master, so it contains the bleeding edge build.
3) If ESLint still can't find the rule, then consider reporting an issue.
-->

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

<!--
Include a ***minimal*** reproduction case.
The more irrelevant code/config you give, the harder it is for us to investigate.

Please consider creating an isolated reproduction repo to make it easy for the volunteer maintainers debug your issue.
-->

```JSON
{
  "rules": {
    "@typescript-eslint/naming-convention": [
      "error",
      {
        "selector": "memberLike",
        "format": ["camelCase"]
      },
      {
        "selector": "property",
        "format": ["camelCase", "PascalCase"]
      }
    ]
  }
}
```

```TS
const foo = { MyProperty: 42 };
```

<!--
Also include your tsconfig, if you're using type-aware linting
-->

**Expected Result**
Because typescript-eslint 4.8.2 doesn't report error for this, I expect typescript-eslint 4.9.x also doesn't report error.

<!--
What did you expect to happen?
Please be specific here - list the exact lines and messages you expect.
-->

**Actual Result**
The following error is reported.
```
  1:15  error  Object Literal Property name `MyProperty` must match one of the following formats: camelCase  @typescript-eslint/naming-convention
```

Previously `property` had a higher precedence than `memberLike`, but it is now a meta selector and the precedence seems to be lower (ref: https://github.com/typescript-eslint/typescript-eslint/pull/2807).

I know it works as expected if I use `"selector": ["classProperty", "objectLiteralProperty", "typeProperty"]` instead of `"selector": "property"`, but I report this issue because it feels like a breaking change in a minor upgrade for me.

<!--
What actually happened?
Please be specific here - list the exact lines and messages that caused errors
-->

**Additional Info**

<!--
Did eslint throw an exception?

Please run your lint again with the --debug flag, and dump the output below.
i.e. eslint --ext ".ts,.js" src --debug
-->

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.9.1` |
| `@typescript-eslint/parser`        | `4.9.1` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `7.15.0` |
| `node`                             | `14.6.0` |

--- a/packages/eslint-plugin/src/rules/naming-convention-utils/shared.ts
+++ b/packages/eslint-plugin/src/rules/naming-convention-utils/shared.ts
@@ -17,4 +17,16 @@ function isMetaSelector(
   return selector in MetaSelectors;
 }
 
-export { selectorTypeToMessageString, isMetaSelector };
+function isMethodOrPropertySelector(
+  selector: IndividualAndMetaSelectorsString | Selectors | MetaSelectors,
+): boolean {
+  return (
+    selector === MetaSelectors.method || selector === MetaSelectors.property
+  );
+}
+
+export {
+  selectorTypeToMessageString,
+  isMetaSelector,
+  isMethodOrPropertySelector,
+};

--- a/packages/eslint-plugin/src/rules/naming-convention-utils/validator.ts
+++ b/packages/eslint-plugin/src/rules/naming-convention-utils/validator.ts
@@ -13,7 +13,11 @@ import {
   UnderscoreOptions,
 } from './enums';
 import { PredefinedFormatToCheckFunction } from './format';
-import { isMetaSelector, selectorTypeToMessageString } from './shared';
+import {
+  isMetaSelector,
+  isMethodOrPropertySelector,
+  selectorTypeToMessageString,
+} from './shared';
 import type { Context, NormalizedSelector } from './types';
 import * as util from '../../util';
 
@@ -47,6 +51,14 @@ function createValidator(
       }
       if (!aIsMeta && bIsMeta) {
         return -1;
+      }
+
+      // for backward compatibility, method and property have higher precedence than other meta selectors
+      if (isMethodOrPropertySelector(a.selector)) {
+        return -1;
+      }
+      if (isMethodOrPropertySelector(b.selector)) {
+        return 1;
       }
 
       // both aren't meta selectors

--- a/packages/eslint-plugin/src/rules/naming-convention-utils/validator.ts
+++ b/packages/eslint-plugin/src/rules/naming-convention-utils/validator.ts
@@ -53,11 +53,14 @@ function createValidator(
         return -1;
       }
 
+      const aIsMethodOrProperty = isMethodOrPropertySelector(a.selector);
+      const bIsMethodOrProperty = isMethodOrPropertySelector(b.selector);
+
       // for backward compatibility, method and property have higher precedence than other meta selectors
-      if (isMethodOrPropertySelector(a.selector)) {
+      if (aIsMethodOrProperty && !bIsMethodOrProperty) {
         return -1;
       }
-      if (isMethodOrPropertySelector(b.selector)) {
+      if (!aIsMethodOrProperty && bIsMethodOrProperty) {
         return 1;
       }
 

--- a/packages/eslint-plugin/tests/rules/naming-convention.test.ts
+++ b/packages/eslint-plugin/tests/rules/naming-convention.test.ts
@@ -1453,6 +1453,31 @@ ruleTester.run('naming-convention', rule, {
         },
       ],
     },
+    {
+      code: `
+        const obj = {
+          Foo: 42,
+          Bar() {
+            return 42;
+          },
+        };
+      `,
+      parserOptions,
+      options: [
+        {
+          selector: 'memberLike',
+          format: ['camelCase'],
+        },
+        {
+          selector: 'property',
+          format: ['PascalCase'],
+        },
+        {
+          selector: 'method',
+          format: ['PascalCase'],
+        },
+      ],
+    },
   ],
   invalid: [
     {

